v1/cloudevents/transport/pubsub/internal/connection_test.go

```diff
@@ -63,26 +153,258 @@ func TestPublishCreateTopic(t *testing.T) {
+				AllowCreateSubscription: true,
+				AllowCreateTopic:        true,
+				Client:                  client,
+				ProjectID:               projectID,
+				TopicID:                 topicID,
+				SubscriptionID:          subID,
+			}
+
+			msg := &pubsub.Message{
+				ID:   "msg-id-1",
+				Data: []byte("msg-data-1"),
+			}
+			// Fails due to injected failure
+			if _, err := psconn.Publish(ctx, msg); err == nil {
+				t.Errorf("Expected publish failure, but didn't see it: %v", err)
+			}
+			// Succeeds
+			if _, err := psconn.Publish(ctx, msg); err != nil {
+				t.Errorf("failed to publish message: %v", err)
+			}
+			verifyTopicDeleteWorks(t, client, psconn, topicID)
+		})
 	}
+}
 
-	if ok, err := client.Topic(topicID).Exists(ctx); err != nil || ok {
+// Test Publishing after Deleting a first version of a topic
+func TestPublishCreateTopicAfterDelete(t *testing.T) {
+	ctx := context.Background()
+	pc := &testPubsubClient{}
+	defer pc.Close()
+
+	projectID, topicID, subID := "test-project", "test-topic", "test-sub"
+
+	client, err := pc.New(ctx, projectID, nil)
+	if err != nil {
+		t.Fatalf("failed to create pubsub client: %v", err)
+	}
+	defer client.Close()
+
+	psconn := &Connection{
+		AllowCreateSubscription: true,
+		AllowCreateTopic:        true,
+		Client:                  client,
+		ProjectID:               projectID,
+		TopicID:                 topicID,
+		SubscriptionID:          subID,
+	}
+
+	msg := &pubsub.Message{
+		ID:   "msg-id-1",
+		Data: []byte("msg-data-1"),
+	}
+	if _, err := psconn.Publish(ctx, msg); err != nil {
+		t.Errorf("failed to publish message: %v", err)
+	}
+
+	verifyTopicDeleteWorks(t, client, psconn, topicID)
+
+	if _, err := psconn.Publish(ctx, msg); err != nil {
+		t.Errorf("failed to publish message: %v", err)
+	}
+
+	verifyTopicDeleteWorks(t, client, psconn, topicID)
+}
+
+// Test that publishing fails if a toic doesn't exist and topic creation isn't allowed
+func TestPublishCreateTopicNotAllowedFails(t *testing.T) {
+	ctx := context.Background()
+	pc := &testPubsubClient{}
+	defer pc.Close()
+
+	projectID, topicID, subID := "test-project", "test-topic", "test-sub"
+
+	client, err := pc.New(ctx, projectID, nil)
+	if err != nil {
+		t.Fatalf("failed to create pubsub client: %v", err)
+	}
+	defer client.Close()
+
+	psconn := &Connection{
+		AllowCreateSubscription: true,
+		AllowCreateTopic:        false,
+		Client:                  client,
+		ProjectID:               projectID,
+		TopicID:                 topicID,
+		SubscriptionID:          subID,
+	}
+
+	msg := &pubsub.Message{
+		ID:   "msg-id-1",
+		Data: []byte("msg-data-1"),
+	}
+	if _, err := psconn.Publish(ctx, msg); err == nil {
+		t.Errorf("publish succeeded unexpectedly")
+	}
+
+	if ok, err := client.Topic(topicID).Exists(ctx); err == nil && ok {
 		t.Errorf("topic id=%s got exists=%v want=false, err=%v", topicID, ok, err)
 	}
 }
 
+// Test that failures of racing topic opens are reported out
+func TestPublishParallelFailure(t *testing.T) {
```

Comment:
No action needed but this test is almost too fancy for me to understand.

Reply:
It's not my favorite test, but it's moderately important in that it makes sure that two concurrent attempts to open a connection get the same result (ie, we don't block one waiting for the other... We don't want multiple parallel requests to send to try to open their own clients.  This ensures that both get a shared failure message if both try to initialize at the same time.  This is obviously hard to force to happen since it's inherently a timing test.    Here we basically try to make sure that the "first" caller blocks for a long time before failing to force the concurrency, but i don't want the test to always take 30 seconds to run.  The for loop is designed to catch the common case with the fast delays, but avoid spurious "failures" with the retries with longer delays.